File: apps/owe-drahn-api/src/app/game/Player.ts
import { PlayerStats } from "./game.utils";

export type FormattedPlayer = {
    life: number;
    points: number;
    uid?: string;
    username: string;
    rank: number;
};

export class Player {
    isPlayersTurn = false;
    ready = false;
    life = 6;
    choosing = false;
    points = 0;
    connected = false;

    uid?: string; // only set if User is logged in
    rank = 0;

    get stats(): PlayerStats | undefined {
        return this._stats;
    }

    set stats(stats: PlayerStats) {
        this._stats = stats;
        this.rank = this.calculateRank(stats.totalGames);
    }

    private _stats?: PlayerStats; // only set if User is logged in

    constructor(readonly id: string, readonly username: string) {}

    getFormattedPlayer(): FormattedPlayer {
        return {
            life: this.life,
            points: this.points,
            uid: this.uid,
            username: this.username,
            rank: this.rank
        };
    }

    private calculateRank(totalGames: number): number {
        return Math.floor(totalGames / 10) + totalGames;
    }

    toJSON() {
        return {
            id: this.id,
            uid: this.uid,
            connected: this.connected,
            username: this.username,
            ready: this.ready,
            isPlayersTurn: this.isPlayersTurn,
            choosing: this.choosing,
            life: this.life,
            points: this.points,
            rank: this.rank,
            stats: this.stats
        };
    }
}

File: apps/gameserver/games/owe_drahn/player.go
package owe_drahn

import (
	"gameserver/games/owe_drahn/models"
)

type Player struct {
	ID          string              `json:"id"`
	UserID      string              `json:"uid"`
	Name        string              `json:"username"`
	Rank        int                 `json:"rank"` // if the user was logged in, we can show their rank
	Stats       *models.PlayerStats `json:"stats"`
	Life        int                 `json:"life"`
	IsReady     bool                `json:"ready"`
	IsChoosing  bool                `json:"choosing"`
	IsConnected bool                `json:"connected"`
	Balance     float64             `json:"balance"` // total wins/losses
}

func NewPlayer(id string, name string) *Player {
	return &Player{
		ID:      id,
		Name:    name,
		Life:    6,
		IsReady: false,
		Balance: 0,
	}
}

func (p *Player) Reset() {
	p.Life = 6
	p.IsReady = false
	p.IsChoosing = false
}

func (p *Player) SetStats(stats *models.PlayerStats) {
	p.Stats = stats
	p.Rank = calculateRank(stats.TotalGames)
}

func (p *Player) ToFormattedPlayer() *models.FormattedPlayer {
	return &models.FormattedPlayer{
		Life:     p.Life,
		UID:      p.UserID,
		Username: p.Name,
		Rank:     p.Rank,
	}
}

func calculateRank(totalGames int) int {
	return totalGames/10 + totalGames
}
